lib/Sema/TypeCheckDecl.cpp

```diff
@@ -3075,14 +3075,34 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
           .fixItReplace(ED->getExtendedTypeRepr()->getSourceRange(),
                         canExtType->getString());
       } else if (!wasAlreadyInvalid) {
-		// If nothing else applies, fall back to a generic diagnostic.
+        // If nothing else applies, fall back to a generic diagnostic.
         ED->diagnose(diag::non_nominal_extension, extType);
       }
       return;
     }
 
     TC.validateExtension(ED);
 
+    extType = ED->getExtendedType();
+    if (extType && !extType->hasError()) {
+      // The first condition catches syntactic forms like
+      //     protocol A & B { ... } // may be protocols or typealiases
+      // The second condition also looks through typealiases and catches
+      //    typealias T = P1 & P2 // P2 is a refined protocol of P1
+      //    extension T { ... }
+      // However, it is trickier to catch cases like
+      //    typealias T = P2 & P1 // P2 is a refined protocol of P1
+      //    extension T { ... }
+      // so we don't do that here.
+      auto extTypeRepr = ED->getExtendedTypeRepr();
+      auto *extTypeNominal = extType->getAnyNominal();
+      if (isa<CompositionTypeRepr>(extTypeRepr)
+          || (extTypeNominal && extTypeNominal != nominal))
+        ED->diagnose(diag::composition_in_extended_type,
+                     nominal->getDeclaredType())
+          .highlight(extTypeRepr->getSourceRange());
```

Comment:
Can we add a fix-it to use the type that the compiler picked?

Reply:
Any suggestions for the fix-it message? How about "would you like to replace the composition with %0"?

Reply:
Maybe “replace composition type %0 with %1”?

Reply:
No message, just a fix-it directly on the warning. It's always the right thing to do in the case where the warning comes up, right?

Reply:
(Or rather, the "extending %0 instead" serves as the message.)

Reply:
> It's always the right thing to do in the case where the warning comes up, right?

Nope. That's why I didn't have a fix-it earlier. It is unclear if the user wants to extend the more specific type or the less specific type. If I were to hazard a guess, I would guess that they want to extend the more specific type.

We are treating it as if they want to extend whatever came first (and not the most specific type), which is potentially not what they want, and can lead to a perplexing type error (as I mentioned in the earlier error message but I got rid of that).

Reply:
If "right thing to do" == "same dynamic semantics", the yes. If "right thing to do" == "matches user expectation on what code should do", then potentially no.

Reply:
:-/ Good point. Then yeah, it would be a separate note. "did you mean to extend the most-inherited protocol %0?" perhaps?

("most-derived" and "refining" aren't quite user-visible terms. Looking at [the Swift Programming Language book](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html) "inherit" is the only term that's really used. "most-specific" might work too.)